man/modinfo.8.scd
MODINFO(8) "kmod" "modinfo"

# NAME

modinfo - Show information about a Linux Kernel module

# SYNOPSIS

*modinfo* [*-0*] [*-F* _field_] [*-k* _kernel_] _modulename_|_filename_

*modinfo* *-V*

*modinfo* *-h*

# DESCRIPTION

*modinfo* extracts information from the Linux Kernel modules given on the
command line. If the module name is not a filename, then the @MODULE_DIRECTORY@/
_version_ directory is searched, as is also done by *modprobe*(8) when loading
kernel modules.

*modinfo* by default lists each attribute of the module in form _fieldname_ :
_value_, for easy reading. The filename is listed the same way (although it's
not really an attribute).

This version of *modinfo* can understand modules of any Linux Kernel
architecture.

# OPTIONS

*-V*, *--version*
	Print the *modinfo* version.

*-F* _field_, *--field* _field_
	Only print this _field_ value, one per line. This is most useful for
	scripts. Field names are case-insensitive. Common fields (which may not
	be in every module) include author, description, license, parm, depends,
	and alias. There are often multiple parm, alias and depends fields. The
	special _field_ filename lists the filename of the module.

*-b* _basedir_, *--basedir* _basedir_
	Root directory for modules, / by default.

*-k* _kernel_
	Provide information about a kernel other than the running one. This is
	particularly useful for distributions needing to extract information
	from a newly installed (but not yet running) set of kernel modules. For
	example, you wish to find which firmware files are needed by various
	modules in a new kernel for which you must make an initrd/initramfs
	image prior to booting.

*-0*, *--null*
	Use the ASCII zero character to separate _field_ values, instead of a new
	line. This is useful for scripts, since a new line can theoretically
	appear inside a _field_.

*-a* *--author*, *-d* *--description*, *-l* *--license*, *-p* *--parameters*, *-n* *--filename*
	These are shortcuts for the *--field* flag's author, description,
	license, parm and filename arguments, to ease the transition from the
	old modutils *modinfo*.

*-h*, *--help*
	Print the help message and exit.

# SEE ALSO

*modprobe*(8)


NEWS
kmod 34
=======

- Improvements

	- Drop pre-built .ko modules from git - distros/packages will need the
	  linux-headers to be able to run the testsuite. There was limited use
	  of the feature, while linters complained about "source-not-included"
	  or "source-contains-prebuilt-binary".

	- Switch build system to meson: autotools is still supported but slated
	  for removal on next release. This is the transition release to help
	  distros and integrators to move to the new build system. Default options
	  target distros while developers can use the build-dev.ini configuration
	  file.

	- Allow to load decompression libraries on demand: liblzma.so, libz.so
	  and libzstd.so can now be loaded on demand, only when there is
	  such a need. For use during early boot for loading modules, if
	  configured well it means none of these libraries are loaded: the
	  module loading logic via finit_module() will just hand over to kernel
	  the open file descriptor and kernel will use its own decompress routine.

	  If kernel doesn't handle decompression or if the module is compressed
	  with a different algorithm than the one configured in the kernel,
	  libkmod can still open the module by dynamically loading the
	  correspondent library.

	  Tools inspecting the module contents, like modinfo, will load that
	  single decompression library instead all of them.

	  For distros building with meson it's possible to choose the behavior
	  per library. Examples: a) -Ddlopen=all uses dlopen behavior for all
	  the libraries; b) -Ddlopen=xz, will make only xz to be dlopen'ed
	  while other enabled libraries will be linked in at build time.

	  The use of dlopen is annotated in the ELF file by using the ELF
	  Package Metadata spec as documented in
	  https://systemd.io/ELF_PACKAGE_METADATA/. Example:

	  $ dlopen-notes.py libkmod.so
	  # build/libkmod.so
	  [
	    {
	      "feature": "xz",
	      "description": "Support for uncompressing xz-compressed modules",
	      "priority": "recommended",
	      "soname": [
	        "liblzma.so.5"
	      ]
	    }
	  ]

	- Add -m / --moduledir to depmod to override in runtime the module
	  directory that was already possible to set on build time. Document
	  the interaction between the dir options: base, module and output.

	- Better error propagation in libkmod for its internal APIs and libc
	  functions up to the callers.

	- Improve libkmod API documentation by adding new sections, documenting
	  functions previously missing, rewording existing ones, adding version
	  information, cross-referencing, etc.

	- Remove deprecated arguments for depmod: --unresolved-error, --quiet,
	  --root and --map.

	- Remove deprecated arguments for rmmod: -w.

	- Remove deprecated arguments for insmod: -p and -s.

	- Add --syslog and --force for insmod to normalize it with other tools.

	- Add bash, fish and zsh shell-completion for insmod, rmmod and lsmod.

	- Remove depmod_module_directory_override from .pc as the kernel side
	  is not making use of it and will likely not need it.

	- Improve builtin module listing and retrieving information from its
	  modinfo index which reduces the amount of needed syscalls by 90%.

	- Improve zstd decompression by using streaming bufferless mode which
	  reduces the amount of syscalls by 65%.

	- Increase use of pread while parsing ELF and indexes in order to reduce
	  syscalls and improve performance.

	- Improve module sorting in depmod to speedup the use of the
	  modules.order index and support duplicate lines in it.

	- Avoid misaligned memory access while reading module signature in
	  libkmod.

	- Add more documentation for contributing to kmod. New developers are
	  welcome to look at the new README.md and CONTRIBUTING.md files for
	  information on process, coding style, build/installation, etc.

	- Overhaul man pages with multiple clarifications, section rewrites and
	  additional documentation.

	- Drop --with-rootlibdir as it's seldom used and was partially broken.

	- Drop strndupa() and alloca() for increased libc compatibility.

	- Better handling of LFS for increased compatibility with libc.

	- Protect kmod_get_dirname() and kmod_new() against NULL argument.

	- Normalize --version / --help output across all tools.

	- Always include log priority in messages, even when building with debug.

	- Optimize index reading by lazily reading nodes on demand, reducing
	  FILE overhead and reducing code duplication wrt FILE vs mmap
	  implementations, etc.

	- Switch index to pre-order to improve performance in both read and
	  write, meaning faster lookup and faster depmod. Some examples:
	  a) traversing all indexes via configuration dump shows a 9%
	  improvement on Raspberry Pi 2. b) writing the indexes takes 90% less
	  lseek() calls, leading to a performance gain of 13%.

	- Make symlink install locations more similar to what distros are
	  using: by default it installs the kmod binary as bin/kmod and the
	  symlinks are located in e.g. `sbin/depmod -> ../bin/kmod`. Changing
	  the sbin location is sufficient to move the symlinks to the
	  appropriate place, so distros using `--sbin /usr/bin` will have them
	  installed in that directory. This avoids distros having to remove the
	  symlink and add the symlinks by themselves. (meson only)

	- Install configuration directories,
	  /{etc,usr/lib}/{depmod,modprobe}.d/ as part of installation, matching
	  what several distros do during packaging. (meson only)

- Bug fixes

	- Fix testsuite using when using configurable module dir.

	- Fix typos on documentation and source code.

	- Fix out of bound access in multiple places when using long paths,
	  synthetic huge files, or handling memory allocation errors, or
	  inconsistent variable types, particularly on 32b builds.

	- Fix internal array APIs, with better error checking: improve execution on
	  very memory-constrained scenarios or very long paths.

	- Fix absolute path handling in depmod.


tools/modinfo.c
// SPDX-License-Identifier: GPL-2.0-or-later
/*
 * kmod-modinfo - query kernel module information using libkmod.
 *
 * Copyright (C) 2011-2013  ProFUSION embedded systems
 */

#include <errno.h>
#include <getopt.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/utsname.h>

#include <shared/util.h>

#include <libkmod/libkmod.h>

#include "kmod.h"

static char separator = '\n';
static const char *field;

struct param {
	struct param *next;
	const char *name;
	const char *param;
	const char *type;
	int namelen;
	int paramlen;
	int typelen;
};

static int add_param(const char *name, size_t namelen, const char *param, size_t paramlen,
		     const char *type, size_t typelen, struct param **list)
{
	struct param *it;

	if (namelen > INT_MAX || paramlen > INT_MAX || typelen > INT_MAX) {
		return -EINVAL;
	}

	for (it = *list; it != NULL; it = it->next) {
		if (it->namelen == (int)namelen && memcmp(it->name, name, namelen) == 0)
			break;
	}

	if (it == NULL) {
		it = malloc(sizeof(struct param));
		if (it == NULL)
			return -ENOMEM;
		it->next = *list;
		*list = it;
		it->name = name;
		it->namelen = namelen;
		it->param = NULL;
		it->type = NULL;
		it->paramlen = 0;
		it->typelen = 0;
	}

	if (param != NULL) {
		it->param = param;
		it->paramlen = paramlen;
	}

	if (type != NULL) {
		it->type = type;
		it->typelen = typelen;
	}

	return 0;
}

static int process_parm(const char *key, const char *value, struct param **params)
{
	const char *name, *param, *type;
	size_t namelen, paramlen, typelen;
	const char *colon = strchr(value, ':');
	int ret;

	if (colon == NULL) {
		ERR("Found invalid \"%s=%s\": missing ':'\n", key, value);
		return 0;
	}

	name = value;
	namelen = colon - value;
	if (streq(key, "parm")) {
		param = colon + 1;
		paramlen = strlen(param);
		type = NULL;
		typelen = 0;
	} else {
		param = NULL;
		paramlen = 0;
		type = colon + 1;
		typelen = strlen(type);
	}

	ret = add_param(name, namelen, param, paramlen, type, typelen, params);
	if (ret < 0) {
		ERR("Unable to add parameter: %s\n", strerror(-ret));
		return -ENOMEM;
	}

	return 0;
}

static int modinfo_params_do(const struct kmod_list *list)
{
	const struct kmod_list *l;
	struct param *params = NULL;
	int err = 0;

	kmod_list_foreach(l, list) {
		const char *key = kmod_module_info_get_key(l);
		const char *value = kmod_module_info_get_value(l);
		if (!streq(key, "parm") && !streq(key, "parmtype"))
			continue;

		err = process_parm(key, value, &params);
		if (err < 0)
			goto end;
	}

	while (params != NULL) {
		struct param *p = params;
		params = p->next;

		if (p->param == NULL)
			printf("%.*s: (%.*s)%c", p->namelen, p->name, p->typelen, p->type,
			       separator);
		else if (p->type != NULL)
			printf("%.*s:%.*s (%.*s)%c", p->namelen, p->name, p->paramlen,
			       p->param, p->typelen, p->type, separator);
		else
			printf("%.*s:%.*s%c", p->namelen, p->name, p->paramlen, p->param,
			       separator);

		free(p);
	}

end:
	while (params != NULL) {
		void *tmp = params;
		params = params->next;
		free(tmp);
	}

	return err;
}

static int modinfo_do(struct kmod_module *mod)
{
	struct kmod_list *l, *list = NULL;
	struct param *params = NULL;
	int err, is_builtin;
	const char *filename = kmod_module_get_path(mod);

	is_builtin = (filename == NULL);

	if (is_builtin) {
		if (field == NULL)
			printf("%-16s%s%c", "name:", kmod_module_get_name(mod), separator);
		else if (field != NULL && streq(field, "name"))
			printf("%s%c", kmod_module_get_name(mod), separator);
		filename = "(builtin)";
	}

	if (field != NULL && streq(field, "filename")) {
		printf("%s%c", filename, separator);
		return 0;
	} else if (field == NULL) {
		printf("%-16s%s%c", "filename:", filename, separator);
	}

	err = kmod_module_get_info(mod, &list);
	if (err < 0) {
		if (is_builtin && err == -ENOENT) {
			/*
			 * This is an old kernel that does not have a file
			 * with information about built-in modules.
			 */
			return 0;
		}
		ERR("could not get modinfo from '%s': %s\n", kmod_module_get_name(mod),
		    strerror(-err));
		return err;
	}

	if (field != NULL && streq(field, "parm")) {
		err = modinfo_params_do(list);
		goto end;
	}

	kmod_list_foreach(l, list) {
		const char *key = kmod_module_info_get_key(l);
		const char *value = kmod_module_info_get_value(l);

		if (field != NULL) {
			if (streq(field, key)) {
				/* filtered output contains no key, just value */
				printf("%s%c", value, separator);
			}
		} else if (streq(key, "parm") || streq(key, "parmtype")) {
			err = process_parm(key, value, &params);
			if (err < 0)
				goto end;
		} else if (separator == '\0') {
			printf("%s=%s%c", key, value, separator);
		} else {
			size_t keylen = strlen(key);
			if (keylen > 15)
				keylen = 15;
			printf("%s:%-*s%s%c", key, 15 - (int)keylen, "", value, separator);
		}
	}

	if (field != NULL)
		goto end;

	while (params != NULL) {
		struct param *p = params;
		params = p->next;

		if (p->param == NULL)
			printf("%-16s%.*s:%.*s%c", "parm:", p->namelen, p->name,
			       p->typelen, p->type, separator);
		else if (p->type != NULL)
			printf("%-16s%.*s:%.*s (%.*s)%c", "parm:", p->namelen, p->name,
			       p->paramlen, p->param, p->typelen, p->type, separator);
		else
			printf("%-16s%.*s:%.*s%c", "parm:", p->namelen, p->name,
			       p->paramlen, p->param, separator);

		free(p);
	}

end:
	while (params != NULL) {
		void *tmp = params;
		params = params->next;
		free(tmp);
	}
	kmod_module_info_free_list(list);

	return err;
}

static int modinfo_path_do(struct kmod_ctx *ctx, const char *path)
{
	struct kmod_module *mod;
	int err = kmod_module_new_from_path(ctx, path, &mod);
	if (err < 0) {
		ERR("Module file %s not found.\n", path);
		return err;
	}
	err = modinfo_do(mod);
	kmod_module_unref(mod);
	return err;
}

static int modinfo_name_do(struct kmod_ctx *ctx, const char *name)
{
	struct kmod_module *mod = NULL;
	int err;

	err = kmod_module_new_from_name_lookup(ctx, name, &mod);
	if (err < 0 || mod == NULL) {
		ERR("Module name %s not found.\n", name);
		return err < 0 ? err : -ENOENT;
	}

	err = modinfo_do(mod);
	kmod_module_unref(mod);

	return err;
}

static int modinfo_alias_do(struct kmod_ctx *ctx, const char *alias)
{
	struct kmod_list *l, *list = NULL;
	int err = kmod_module_new_from_lookup(ctx, alias, &list);
	if (err < 0) {
		ERR("Module alias %s not found.\n", alias);
		return err;
	}

	if (list == NULL) {
		ERR("Module %s not found.\n", alias);
		return -ENOENT;
	}

	kmod_list_foreach(l, list) {
		struct kmod_module *mod = kmod_module_get_module(l);
		int r = modinfo_do(mod);
		kmod_module_unref(mod);
		if (r < 0)
			err = r;
	}
	kmod_module_unref_list(list);
	return err;
}

static const char cmdopts_s[] = "adlpn0mF:k:b:Vh";
static const struct option cmdopts[] = {
	{ "author", no_argument, 0, 'a' },
	{ "description", no_argument, 0, 'd' },
	{ "license", no_argument, 0, 'l' },
	{ "parameters", no_argument, 0, 'p' },
	{ "filename", no_argument, 0, 'n' },
	{ "null", no_argument, 0, '0' },
	{ "modname", no_argument, 0, 'm' },
	{ "field", required_argument, 0, 'F' },
	{ "set-version", required_argument, 0, 'k' },
	{ "basedir", required_argument, 0, 'b' },
	{ "version", no_argument, 0, 'V' },
	{ "help", no_argument, 0, 'h' },
	{ NULL, 0, 0, 0 },
};

static void help(void)
{
	printf("Usage:\n"
	       "\t%s [options] <modulename|filename> [args]\n"
	       "Options:\n"
	       "\t-a, --author                Print only 'author'\n"
	       "\t-d, --description           Print only 'description'\n"
	       "\t-l, --license               Print only 'license'\n"
	       "\t-p, --parameters            Print only 'parm'\n"
	       "\t-n, --filename              Print only 'filename'\n"
	       "\t-0, --null                  Use \\0 instead of \\n\n"
	       "\t-m, --modname               Handle argument as module name instead of alias or filename\n"
	       "\t-F, --field FIELD           Print only provided FIELD\n"
	       "\t-k, --set-version VERSION   Use VERSION instead of $(uname -r)\n"
	       "\t-b, --basedir DIR           Use DIR as filesystem root for " MODULE_DIRECTORY
	       "\n"
	       "\t-V, --version               Show version\n"
	       "\t-h, --help                  Show this help\n",
	       program_invocation_short_name);
}

static bool is_module_filename(const char *name)
{
	struct stat st;

	if (path_ends_with_kmod_ext(name, strlen(name)) && stat(name, &st) == 0 &&
	    S_ISREG(st.st_mode))
		return true;

	return false;
}

static int do_modinfo(int argc, char *argv[])
{
	struct kmod_ctx *ctx;
	char dirname_buf[PATH_MAX];
	const char *dirname = NULL;
	const char *kversion = NULL;
	const char *root = NULL;
	const char *null_config = NULL;
	bool arg_is_modname = false;
	int i, err;

	for (;;) {
		int c, idx = 0;
		c = getopt_long(argc, argv, cmdopts_s, cmdopts, &idx);
		if (c == -1)
			break;
		switch (c) {
		case 'a':
			field = "author";
			break;
		case 'd':
			field = "description";
			break;
		case 'l':
			field = "license";
			break;
		case 'p':
			field = "parm";
			break;
		case 'n':
			field = "filename";
			break;
		case '0':
			separator = '\0';
			break;
		case 'm':
			arg_is_modname = true;
			break;
		case 'F':
			field = optarg;
			break;
		case 'k':
			kversion = optarg;
			break;
		case 'b':
			root = optarg;
			break;
		case 'h':
			help();
			return EXIT_SUCCESS;
		case 'V':
			kmod_version();
			return EXIT_SUCCESS;
		case '?':
			return EXIT_FAILURE;
		default:
			ERR("unexpected getopt_long() value '%c'.\n", c);
			return EXIT_FAILURE;
		}
	}

	if (optind >= argc) {
		ERR("missing module or filename.\n");
		return EXIT_FAILURE;
	}

	if (root != NULL || kversion != NULL) {
		struct utsname u;
		int n;
		if (root == NULL)
			root = "";
		if (kversion == NULL) {
			if (uname(&u) < 0) {
				ERR("uname() failed: %m\n");
				return EXIT_FAILURE;
			}
			kversion = u.release;
		}

		n = snprintf(dirname_buf, sizeof(dirname_buf),
			     "%s" MODULE_DIRECTORY "/%s", root, kversion);
		if (n >= (int)sizeof(dirname_buf)) {
			ERR("bad directory %s" MODULE_DIRECTORY "/%s: path too long\n",
			    root, kversion);
			return EXIT_FAILURE;
		}
		dirname = dirname_buf;
	}

	ctx = kmod_new(dirname, &null_config);
	if (!ctx) {
		ERR("kmod_new() failed!\n");
		return EXIT_FAILURE;
	}

	err = 0;
	for (i = optind; i < argc; i++) {
		const char *name = argv[i];
		int r;

		if (arg_is_modname)
			r = modinfo_name_do(ctx, name);
		else if (is_module_filename(name))
			r = modinfo_path_do(ctx, name);
		else
			r = modinfo_alias_do(ctx, name);

		if (r < 0)
			err = r;
	}

	kmod_unref(ctx);
	return err >= 0 ? EXIT_SUCCESS : EXIT_FAILURE;
}

const struct kmod_cmd kmod_cmd_compat_modinfo = {
	.name = "modinfo",
	.cmd = do_modinfo,
	.help = "compat modinfo command",
};
